### PR TITLE
Sprint 27

### PR DESCRIPTION
**Pull request by ericsorenson**
# Sprint 27

## [2448 Revisit Message Flow](https://github.com/flexion/ef-cms/issues/2448)
image
## [3407 Add Judge association for each case](https://github.com/flexion/ef-cms/issues/3407)
image
## [2377 Docket Clerk: Ability to manually change case status](https://github.com/flexion/ef-cms/issues/2377)
image
## [2156 Court User: Ability to track pending motions and other items](https://github.com/flexion/ef-cms/issues/2156)
image
## [2380 Docket Clerk: Ability to add court-issued documents to the docket record](https://github.com/flexion/ef-cms/issues/2380)
image

In this PR we made a couple of changes related to how permissions work.

#### AWS Account Specific Roles & Policies

The first main change is that the circle ci policy is no longer managed from a policy.json.tpl file; instead, we run terraform locally to create and update that permission.  Those permissions can be found in the `iam/account-specific/main/circle-ci.tf` file.  In order to create that policy from this PR, and update future changes, you'll need to run terraform locally using the following command: `cd iam/account-specific/terraform/main && ../bin/deploy-app.sh`.  Note that you will need an aws user or role setup locally with permissions to run iam:CreateRole & iam:CreatePolicy, and probably some other permissions.  You can always setup a temporary admin user, run terraform, then delete that admin user if necessary.  After creating this new policy, you'll need to link you current aws user's policy (the one that circle ci uses) with this new policy.  Delete the old policy.

After running this, please keep track of the ARNs printed out to the console.  These will need to be configured as a CircleCI env variable.

#### Environment Specific Roles & Policies

The second change is since the circle policy no longer has permissions to create roles or policies, we have to manually create the policies that are used in various places in our environment.  For example, lambda needs a separate role.  There is a separate role needed for s3 replication.  There is a role needed for cognito triggers, etc.  To setup these various roles and policies in an environment, you will need to run the following:

`cd iam/environment-specific/terraform/main && ../bin/deploy-app stg`

Again, this must be run with a user who has the proper permissions for creating and editing policies and roles.

After running this, please keep track of the ARNs printed out to the console.  These will need to be configured as a CircleCI env variable.

#### New CircleCI Environment Variables
We need to setup environment variables in CircleCI so that our aws resources know their role arns.

Add the following as CircleCI env variables:
- `CLOUDWATCH_ROLE_ARN`
- `POST_CONFIRMATION_ROLE_ARN_DEV` // you can skip since you don't have dev
- `POST_CONFIRMATION_ROLE_ARN_STG`
- `POST_CONFIRMATION_ROLE_ARN_PROD`

### CODE REVIEW

**Comment by waldoj** [comment]
### PR Review Checklist

- [ ] begin by deleting any existing copy of the Git repository and cloning a fresh copy, running `docker system prune` to remove any cached artifacts from any prior code review
- [ ] review the pull request itself, to get oriented
	- [ ] read the description of the pull request, which should summarize the changes made
	- [ ] read through every task on the Scrum board that's encompassed by this pull request
	- [ ] read the description of the commits that comprise the pull request
- [ ] stand up the site locally, with `./docker-run.sh`
	- [ ] test all functionality in all major browsers, emphasizing the functionality that this pull request addresses
		- [ ] for internal Court functionality, perform the most thorough testing in Chrome, though also test in Edge and Firefox
		- [ ] for public-facing functionality, test in browsers consistent with [public browser use data](https://analytics.usa.gov/), including Internet Explorer 11
		- [ ] test in Mobile Safari and Mobile Chrome, with the caveat that not all internal Court functionality will be necessary on these platforms
	- [ ] use an automated audit tool for code quality and practices (recommended: [Chrome DevTools](https://developers.google.com/web/tools/chrome-devtools/), aka [Lighthouse](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk))
		- [ ] look at efficiency of page loads, asset sizes, HTTP connection management, etc.
	- [ ] review for accessibility
		- [ ] use an automated audit tool, such as Chrome Audit or aXe
		- [ ] navigate site only with the keyboard
		- [ ] use VoiceOver or Narrator to navigate the site with audio only, with the display turned off
		- [ ] manually test anything that pa11y cannot test automatically (e.g., contrast of text over images)
- [ ] review static code analysis results in SonarCloud
- [ ] examine OWASP ZAP output in `docs/`, to ensure that any errors are known to be false positives or have been previously declared to be acceptable
- [ ] skim all new code, in the context of existing code, [looking for problems](#what-we-look-for) (knowing that the vast majority of new code will be covered by tests)
- [ ] review all tests
	- [ ] look at code coverage of tests in SonarCloud
	- [ ] methodically review all new tests for correctness, quality of naming
	- [ ] determine what code isn’t tested, review that rigorously
- [ ] review documentation to ensure that it matches changes
- [ ] provide comments on the pull request on GitHub, as necessary
	- [ ] for comments that are specific to a particular line of code, comment on those specific lines
	- [ ] for comments that are more general, attach the comment to a random line in `README.md` (as opposed to commenting on the pull request itself), to be able to use GitHub's ability to thread discussions on those comments
- [ ] for each feature-level bug (i.e., it’s working as designed, but designed wrong), open a new issue and put it in the backlog

**Comment by waldoj** [comment]
Because GitHub lacks effective pull-request-level feedback mechanisms that must be addressed before a pull request can be merged, I'll be using [`web-api/README.md`](https://github.com/ustaxcourt/ef-cms/pull/113/files#diff-ded5ed3a95f825c4f38e902cdcf3c62c) as the place to attach PR-level comments.

File: web-api/run-serverless.sh [hunk 1]
```diff
@@ -18,11 +18,13 @@ ACCOUNT_ID="${ACCOUNT_ID#\"}"
 export NODE_PRESERVE_SYMLINKS=1
 find ./web-api/src -type f -exec chmod -R ugo+r {} ";"
 
+npm run build:assets
 npm run "${build}"
 cp "./web-api/src/${handler}" /tmp
 cp "./dist/${handler}" web-api/src
 
 export SLS_DEPLOYMENT_BUCKET="${EFCMS_DOMAIN}.efcms.${slsStage}.${region}.deploys"
+export SLS_DEBUG=*
```
Comment (waldoj) [inline web-api/run-serverless.sh]:
```suggestion
export SLS_DEBUG="*"
```
Shouldn't this be quoted?

File: web-api/run-umzug.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash -e
+ENV=$1
+node ./web-api/umzug.js $ENV
```
Comment (waldoj) [inline web-api/run-umzug.sh]:
```suggestion
node ./web-api/umzug.js "$ENV"
```
I'm pretty sure this should be quoted.

File: web-api/README.md [hunk 3]
```diff
@@ -22,20 +22,10 @@ This will host a local service at http://localhost:3000.
 
 # Trouble Shooting
 
-Sometimes you may get errors when doing npm install.  To fix, removce the package-lock.json file and re-run npm install.
+Sometimes you may get errors when doing npm install.  To fix, remove the package-lock.json file and re-run npm install.
```
[the remark this answers is not in the document]
Comment (ericsorenson) [reply web-api/README.md]: Updated!

**Comment by waldoj** [comment]
@JessicaMarine, I have reviewed this pull request, and I now recommend that you merge it to `staging`.
